### PR TITLE
bitratechecker

### PR DESCRIPTION
**Pull request by arcctgx**
The purpose of this tool is to guess if a set of `.mp3` files originates from the same source.

In the early days of Internet music downloads (times of Napster, Audiogalaxy or WinMX) it was common to download an album where individual tracks came from different sources: encoded using different software, with different bitrates and different bitrate modes. It's not as common today, but still happens.

This tool detects such inconsistent sets of `.mp3` files. The information about bitrate value, bitrate mode and encoder library string are gathered using `mediainfo` (columns 2-4). Further detection of an encoder is done by `mp3guessenc` (column 5). If all data are the same for all files (with the exception of bitrate values if all files are VBR), then the set is considered uniform.

The same information `mediainfo` provides can be obtained with `Mutagen` library - but that will be the next step.

Example output:
```
Checking music/downloads/low/long_division - OK
Checking music/downloads/low/things_we_lost_in_the_fire - NOK
01_sunflower.mp3                    192000  CBR             unknown                 FhG (ACM or producer pro)
02_whitetail.mp3                    192000  CBR             unknown                 FhG (ACM or producer pro)
03_dinosaur_act.mp3                 128000  CBR             unknown                 FhG (l3enc, fastenc or mp3enc)
04_medicine_magazines.mp3           191869  VBR             LAME3.86                Lame
05_laser_beam.mp3                   160000  CBR             unknown                 FhG (l3enc, fastenc or mp3enc)
06_july.mp3                         192728  VBR             LAME3.86                Lame
07_embrace.mp3                      192000  CBR             unknown                 FhG (ACM or producer pro)
08_whore.mp3                        192844  VBR             LAME3.86                Lame
09_kind_of_girl.mp3                 193002  VBR             LAME3.86                Lame
10_like_a_forest.mp3                196116  VBR             LAME3.86                Lame
11_closer.mp3                       191748  VBR             LAME3.86                Lame
12_untitled.mp3                     192000  CBR             unknown                 FhG (ACM or producer pro)
13_in_metal.mp3                     192932  VBR             LAME3.86                Lame
```

### CODE REVIEW

File: bitratechecker.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python2
+
+import sys
+import os
+import glob
+import subprocess
+import json
+
+def usage():
+    print "usage: " + os.path.basename(sys.argv[0]) + " <dir> [dir2] ..."
```
Comment (BartoszCiesla) [inline bitratechecker.py]: I know it is python2 - at least shebang says so but I encourage to use py3 syntax - it will be easier to migrate later on - `print("...")`

File: bitratechecker.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+        encoder.append(enc)
+
+    return encoder
+
+
+if len(sys.argv) < 2:
+    usage()
+
+for path in sys.argv[1:]:
```
Comment (BartoszCiesla) [inline bitratechecker.py]: Even for small programs I used to use `OptParse` and recently `ArgParse` - provides automatically features like usage printing, help, etc.

**Review by BartoszCiesla — APPROVED** [review]

File: bitratechecker.py (hunk 1 above)
Comment (arcctgx) [reply bitratechecker.py]: I think being used to Gerrit is a liability here on GitHub :rofl:
The stack of commits is displayed from oldest to newest - newest is at the bottom.
I think you were looking at the wrong commit - please see the state at 5c1d722. It uses Py3-style printing.

File: bitratechecker.py (hunk 2 above)
Comment (arcctgx) [reply bitratechecker.py]: Please see the state at 5c1d722 where `argparse` is used.

File: bitratechecker.py (hunk 1 above)
Comment (arcctgx) [reply bitratechecker.py]: If you click on "Files changed" at the top you'll see the final state of all files at 5c1d722.

Comment (BartoszCiesla) [reply bitratechecker.py]: Figured out that after going through all commits ;-)

File: bitratechecker.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                enc = subprocess.check_output(["mp3guessenc", "-n", f]).splitlines()[0]
+            except subprocess.CalledProcessError as e:
+                # mp3guessenc returns different statuses for different encoders - not just zero on success
+                # so we actually get meaningful output here in except block
+                enc = e.output.splitlines()[0]
+
+            self.__encoder.append(enc)
+
+
+    def print_formatted(self):
```
Comment (arcctgx) [inline bitratechecker.py]: @BCiesla maybe it would be better to implement `__str__` instead of `print_formatted`?

File: bitratechecker.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            print("NOK")
+            download.print_formatted()
+        else:
+            print("OK")
+            if args.verbose:
+                download.print_formatted()
+    except ValueError:
+        print("no .mp3 files found!")
+    except EnvironmentError:
+        print("failed to get metadata using mediainfo!")
```
Comment (arcctgx) [inline bitratechecker.py]: @BCiesla I was considering to have class definition and this part of "real" program that uses the class in separate files - or is it overkill?

File: bitratechecker.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                enc = e.output.splitlines()[0]
+
+            self.__encoder.append(enc)
+
+
+    def print_formatted(self):
+        s = []
+
+        for z in zip(self.__name, self.__bitrate, self.__mode, self.__library, self.__encoder):
+            s.append("{:<50}\t{}\t{:<12}\t{:<20}\t{}".format(z[0], z[1], z[2], z[3], z[4]))
```
Comment (arcctgx) [inline bitratechecker.py:79]: ...and in Python3 I could use `*z` instead of this ugly `z[0], z[1], ...` :frowning_face:

File: bitratechecker.py (hunk 3 above)
Comment (BartoszCiesla) [reply bitratechecker.py]: Yes, that's good idea

File: bitratechecker.py (hunk 4 above)
Comment (BartoszCiesla) [reply bitratechecker.py]: IMHO it depends - that's a relatively small program and I usually keep it like this but if that's going to grow or class will be used elsewhere then separate module is way to go.

File: bitratechecker.py (hunk 5 above)
Comment (BartoszCiesla) [reply bitratechecker.py:79]: I'm not a big fan of this expression but you're right ;-)

File: bitratechecker.py (hunk 4 above)
Comment (arcctgx) [reply bitratechecker.py]: I doubt it's going to grow significantly. If anything, ditching `mediainfo` and switching to `Mutagen` might make it smaller. And the class itself is quite specialized, so I don't know if it will be used in other context. I'll leave it as it is now, it can always be extracted if it becomes necessary.

File: bitratechecker.py (hunk 5 above)
Comment (arcctgx) [reply bitratechecker.py:79]: I'm becoming a fan of concise syntax.

File: bitratechecker.py (hunk 3 above)
Comment (arcctgx) [reply bitratechecker.py]: Done.

File: bitratechecker.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python2
+
+from __future__ import print_function
+import sys
+import os
+import glob
+import subprocess
+import json
+import argparse
```
Comment (skaluzka) [inline bitratechecker.py]: Personally I prefer sorted order (nice discussion about that: https://stackoverflow.com/a/20763446).

File: bitratechecker.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+class Mp3DownloadAnalyzer(object):
+    def __init__(self, mp3_files):
+
+        if not len(mp3_files) > 0:
+            raise ValueError
```
Comment (skaluzka) [inline bitratechecker.py]: Please consider something like that:
```
def __init_(self, mp3_files):
    self.mp3_files = mp3_files

    if not len(self.mp3_files) > 0:
    ...
```

you could then remove ``mp3_files`` from ``__get_media_info()`` and ``__guess_encoder()`` and use ``self.mp3_files`` inside.

File: bitratechecker.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+
+    def __str__(self):
+        s = []
+
+        for z in zip(self.__name, self.__bitrate, self.__mode, self.__library, self.__encoder):
+            s.append("{:<50}\t{}\t{:<12}\t{:<20}\t{}".format(z[0], z[1], z[2], z[3], z[4]))
+        s.append("")
+
+        return "\n".join(s)
```
Comment (skaluzka) [inline bitratechecker.py:82]: Nice, but I would like to see tests for that or at least simple example embedded in docstrings ;)

File: bitratechecker.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+            return False
+
+        if not self.__all_cbr and not self.__all_vbr:
+            return False
+
+        if self.__all_cbr and not self.__same_bitrate:
+            return False
+
+        return True
+
```
Comment (skaluzka) [inline bitratechecker.py:99]: Personally I prefer clear indication of so-called entry-point in every script. For more detail please check below links:
https://docs.python.org/3/library/__main__.html
https://docs.python.org/2.7/library/__main__.html

File: bitratechecker.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+            (default: current directory)""")
+args = parser.parse_args()
+
+for path in args.directory:
+    if not os.path.isdir(path):
+        continue
+
+    mp3_files = sorted(glob.glob(os.path.join(path, "*.[mM][pP]3")))
+
+    print("Checking", path, "-", end=' ')
```
Comment (skaluzka) [inline bitratechecker.py]: If I see correctly here is probably inconsistent prints format...
```
print("Checking" ...)
...
print("NOK")
...
print("OK")
...
print("no .mp3 files found!")
...
print("failed to get metadata using mediainfo!")
```

I think you are mixing upper and lower cases here, but if this is intended then I am wrong (sorry).

File: bitratechecker.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+    try:
+        download = Mp3DownloadAnalyzer(mp3_files)
```
Comment (skaluzka) [inline bitratechecker.py]: Personally I would use ``downloader`` name here :)

File: bitratechecker.py (hunk 6 above)
Comment (arcctgx) [reply bitratechecker.py]: I think current order reflects how these imports were added during development. I don't have a strong opinion about that, if there are reasons for sorting the imports I will do that.

File: bitratechecker.py (hunk 10 above)
Comment (arcctgx) [reply bitratechecker.py]: Actually that's the way I wanted it to be, this is what it looks like in practice:
```
Checking qqq/wwwww/ - failed to get metadata using mediainfo!
Checking aaa/bbb/ - OK
Checking xxxxxxx/yyy/ - no .mp3 files found!
Checking zzzzz/xxx/ - NOK
(...)
```

File: bitratechecker.py (hunk 11 above)
Comment (arcctgx) [reply bitratechecker.py]: Are you referring to the name of instance, or class?
For the instance `download` makes more sense to me - a "download" (as a noun) is a set of downloaded files.

File: bitratechecker.py (hunk 7 above)
Comment (arcctgx) [reply bitratechecker.py]: Good idea. I think `mp3_files` was still used as argument because it was an artifact of an early version, when there were two separate functions without a class.

File: bitratechecker.py (hunk 9 above)
Comment (arcctgx) [reply bitratechecker.py:99]: Yes, I thought about it but didn't have enough motivation. :slightly_smiling_face:
I created a separate funtion for parsing command-line arguments, and `main()` that is only called when program is executed directly (not as an import).

File: bitratechecker.py (hunk 8 above)
Comment (arcctgx) [reply bitratechecker.py:82]: I don't intend to write any automated tests now. I tested this quite extensively already (and identified many edge cases in early stage of development).

Regarding docstring - do I understand correctly you're suggesting to embed an example similar to the one in description of this PR?
